float32 left
float32 right
float32 mid

float32 angle